recipes_source/mobile_perf.rst

```diff
@@ -0,0 +1,191 @@
+Pytorch Mobile Performance Recipes
+==================================
+
+Introduction
+------------
+Performance (aka latency) is crucial to most, if not all,
+applications and use-cases of ML model inference on mobile devices.
+
+Today, PyTorch executes the models on the CPU backend pending availability
+of other hardware backends such as GPU, DSP, and NPU.
+
+Model preparation
+-----------------
+
+Next recipes you can take (offline) while preparing the model
+to have an optimized model that will probably have shorter execution time
+(higher performance, lower latency) on the mobile device.
+
+1. Use torch.utils.mobile_optimizer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Torch mobile_optimizer package does several optimizations with the model,
+which will help to conv2d and linear operations.
+It pre-packs model weights in an optimized format and fuses ops above with relu
+if it is the next operation.
+
+::
+
+  from torch.utils.mobile_optimizer import optimize_for_mobile
+  traced_model = torch.jit.load("input_model_path")
+  optimized_model = optimize_for_mobile(traced_model)
+  torch.jit.save(optimized_model, "output_model_path")
+
+2. Fuse operators using ``torch.quantization.fuse_modules``
+Do not be confused that fuse_modules is in the quantization package.
+It works for all types of torch script modules.
+``torch.quantization.fuse_modules`` fuses a list of modules into a single module.
+It fuses only the following sequence of modules:
+
+- Convolution, Batch normalization
+- Convolution, Batch normalization, Relu
+- Convolution, Relu
+- Linear, Relu
+
+This script shows how to fuse Batch Normalization in torchvision mobileNetV2.
+
+::
+
+  import torch
+  import torchvision
+
+  # ConvBNReLU: nn.Sequential[nn.Conv2d, nn.BatchNorm2d, nn.ReLU]
+  # InvertedResidual: nn.Sequential[ConvBnReLU, ConvBNReLU, nn.Conv2d, nn.BatchNorm2d]
+  def fuse_model(model):
+      for idx, m in enumerate(model.modules()):
+          name = m._get_name();
+          if name == 'ConvBNReLU':
+              torch.quantization.fuse_modules(m, ['0', '1'], inplace=True)
+          if name == 'InvertedResidual':
+              for idx in range(len(m.conv)):
+                  if type(m.conv[idx]) == torch.nn.Conv2d:
+                      torch.quantization.fuse_modules(m.conv, [str(idx), str(idx + 1)], inplace=True)
+
+  m = torchvision.models.mobilenet_v2(pretrained=True)
+  m.eval()
+  fuse_model(m)
+  torch.jit.trace(m, torch.rand(1, 3, 224, 224)).save("mobilenetV2-bnfused.pt")
```

Comment:
Should we be guiding people to `torch.jit.trace()`? `torch.jit.script()` preserves control flow, `trace()` does not. `trace()` is still there for cases where `script()` hits an unsupported op.